### PR TITLE
Added pre-migration infra and a pre-migration for fixing schema ID migration numbers

### PR DESCRIPTION
**Pull request by harshilsharma63**
#### Summary
1. Added code infra for running pre-migrations. Pre migrations are run just before running Morph based DB migrations.
2. Added the first pre migration for automated fix for https://github.com/mattermost/mattermost/pull/36866

#### Ticket Link
Fixes https://mattermost.atlassian.net/browse/MM-68848

#### Release Note
```release-note
Added a pre-migration setup to fix the incorrect database migration numbers that prevented upgrading Mattermost from v10.11 to v11.7
```

## Change Impact: 🔴 High

**Regression Risk:** Changes modify migration orchestration and perform in-place updates to db_migrations at startup, directly affecting data integrity and upgrade paths; while new unit tests cover helper paths, runtime startup behavior, multi-node upgrades, and some table-existence branches remain risky.  
**QA Recommendation:** Strong manual QA and integration testing required—exercise fresh installs and the v10.11 ESR → v11.7 ESR upgrade path, verify renumbering mappings and idempotency, and test multi-node upgrade, failure/recovery, and rollback scenarios before rollout.

*Generated by CodeRabbitAI*

### CODE REVIEW

File: server/channels/store/sqlstore/store.go [hunk 1]
```diff
@@ -1138,3 +1143,126 @@ func (ss *SqlStore) ScheduledPost() store.ScheduledPostStore {
...
+		handler        func() error
+	}
+
+	migrations := []sqlMigration{
+		{"renumber_roles_schemeid_migrations", 144, ss.doRenumberRolesSchemeIdMigrations},
+	}
+
+	if len(migrations) == 0 {
+		return nil
+	}
```
Comment (agarciamontoro) [inline server/channels/store/sqlstore/store.go]: We can remove this.

File: server/channels/store/sqlstore/store.go [hunk 2]
```diff
@@ -1138,3 +1143,126 @@ func (ss *SqlStore) ScheduledPost() store.ScheduledPostStore {
...
+	if _, err := ss.GetMaster().Exec("UPDATE db_migrations SET Version = 156 WHERE Version = 142 AND Name = 'add_schemeid_to_roles'"); err != nil {
+		return errors.Wrap(err, "failed to renumber add_schemeid_to_roles")
+	}
+	if _, err := ss.GetMaster().Exec("UPDATE db_migrations SET Version = 157 WHERE Version = 143 AND Name = 'backfill_roles_schemeid'"); err != nil {
+		return errors.Wrap(err, "failed to renumber backfill_roles_schemeid")
+	}
+	if _, err := ss.GetMaster().Exec("UPDATE db_migrations SET Version = 158 WHERE Version = 144 AND Name = 'add_roles_schemeid_index'"); err != nil {
+		return errors.Wrap(err, "failed to renumber add_roles_schemeid_index")
+	}
+	return nil
```
Comment (agarciamontoro) [inline server/channels/store/sqlstore/store.go:1242]: Can we add a comment here mentioning what the problem was and why we're doing this?

File: server/channels/store/sqlstore/store.go [hunk 3]
```diff
@@ -1138,3 +1143,126 @@ func (ss *SqlStore) ScheduledPost() store.ScheduledPostStore {
...
+	}
+
+	for _, m := range migrations {
+		applied, err := ss.isDBMigrationApplied(m.minDBMigration)
+		if err != nil {
+			return errors.Wrapf(err, "failed to check DB migration %d status", m.minDBMigration)
+		}
+		if !applied {
+			continue
+		}
```
Comment (agarciamontoro) [inline server/channels/store/sqlstore/store.go:1193]: Do we need this check? As far as I understand, it's not needed for the current pre-migration, since their `UPDATE`s are no-ops if the migrations do not exist. I would remove this to simplify this as much as we can.

**Review by agarciamontoro — CHANGES_REQUESTED** [review]
The logic seems fine to me, thank you! Just a couple of comments below.

We'd need to test this well before merging, but I've been thinking of edge cases and I couldn't come up with any faulty counter-example.

File: server/channels/store/sqlstore/store.go [hunk 4]
```diff
@@ -1138,3 +1143,126 @@ func (ss *SqlStore) ScheduledPost() store.ScheduledPostStore {
...
+		if err := ss.markPreMigrationComplete(m.name); err != nil {
+			return errors.Wrapf(err, "failed to mark pre-migration %q complete", m.name)
+		}
+		ss.logger.Debug("Completed pre-migration", mlog.String("name", m.name))
+	}
+
+	return nil
+}
+
+func (ss *SqlStore) doRenumberRolesSchemeIdMigrations() error {
```
Comment (isacikgoz) [inline server/channels/store/sqlstore/store.go:1224]: I think we should run these in one transaction

File: server/channels/store/sqlstore/store.go [hunk 5]
```diff
@@ -1138,3 +1143,126 @@ func (ss *SqlStore) ScheduledPost() store.ScheduledPostStore {
...
+
+	if len(migrations) == 0 {
+		return nil
+	}
+
+	// preMigration runs before the schema migrations, so on a fresh install the
+	// Systems table (created by schema migration 000015) does not yet exist.
+	// In that case there is no prior data to migrate, so we skip. Systems
+	// existing also implies db_migrations exists (Morph created it).
+	exists, err := ss.systemsTableExists()
```
Comment (isacikgoz) [inline server/channels/store/sqlstore/store.go]: Why do you check the systems table? I think we should focus on whether `db_migrations` table exists before migrations. And I think it's not. We should also check that one, `isDBMigrationApplied` may result in a false positive case.

File: server/channels/store/sqlstore/store.go [hunk 6]
```diff
@@ -1138,3 +1143,126 @@ func (ss *SqlStore) ScheduledPost() store.ScheduledPostStore {
...
+	if errors.Is(err, sql.ErrNoRows) {
+		return false, nil
+	}
+	if err != nil {
+		return false, errors.Wrap(err, "unable to select from Systems")
+	}
+	return value == "true", nil
+}
+
+func (ss *SqlStore) markPreMigrationComplete(name string) error {
```
Comment (isacikgoz) [inline server/channels/store/sqlstore/store.go:1283]: We don't need to record this with the app migration, `db_migrations` table should be the only place we track DB schema state.

**Review by isacikgoz — CHANGES_REQUESTED** [review]

File: server/channels/store/sqlstore/store.go (hunk 5 above)
Comment (harshilsharma63) [reply server/channels/store/sqlstore/store.go]: I believe I need to check for both tables. `db_migrations` check is needed to detect an empty DB and skip premigrations, and `systems` table because thats where I'm storing whether each individual pre migration has run or not, so as to not run them each time, very similar to app migrations.

File: server/channels/store/sqlstore/store.go (hunk 6 above)
Comment (harshilsharma63) [reply server/channels/store/sqlstore/store.go:1283]: This is just like app migrations where we store the status in systems table to not run them each time.

File: server/channels/store/sqlstore/store.go (hunk 4 above)
Comment (harshilsharma63) [reply server/channels/store/sqlstore/store.go:1224]: Done.

File: server/channels/store/sqlstore/store.go (hunk 3 above)
Comment (harshilsharma63) [reply server/channels/store/sqlstore/store.go:1193]: Its not needed for the current one but I added a generic system to add arbitrary pre-migrations, similar to app migrations.

File: server/channels/store/sqlstore/store.go (hunk 2 above)
Comment (harshilsharma63) [reply server/channels/store/sqlstore/store.go:1242]: Done.

File: server/channels/store/sqlstore/store.go (hunk 1 above)
Comment (harshilsharma63) [reply server/channels/store/sqlstore/store.go]: Makes sense. Done.

**Comment by harshilsharma63** [comment]
@coderabbitai full review

File: server/channels/store/sqlstore/store.go [hunk 7]
```diff
@@ -1138,3 +1143,143 @@ func (ss *SqlStore) ScheduledPost() store.ScheduledPostStore {
...
+}
+
+func (ss *SqlStore) tableExists(tableName string) (bool, error) {
+	var exists bool
+	if err := ss.GetMaster().Get(&exists, `
+		SELECT EXISTS (
+			SELECT 1 FROM information_schema.tables
+			WHERE LOWER(table_name) = $1
+		)
+	`, tableName); err != nil {
```
[the remark this answers is not in the document]
Comment (harshilsharma63) [reply server/channels/store/sqlstore/store.go]: Good catch. This is now fixed.

File: server/channels/store/sqlstore/store.go [hunk 8]
```diff
@@ -233,6 +233,11 @@ func New(settings model.SqlSettings, logger mlog.LoggerIFace, metrics einterface
 	}
 
 	if !store.skipMigrations {
+		err = store.preMigration()
+		if err != nil {
+			return nil, errors.Wrap(err, "error while running pre-migrations")
+		}
```
[the remark this answers is not in the document]
Comment (harshilsharma63) [reply server/channels/store/sqlstore/store.go:239]: This is not a problem. There are multiple existing error paths where the caller doesn't receive the store.

**Comment by harshilsharma63** [comment]
@lindalumitchell heads up will need testing on this one, especially the upgrade path from v10.11.17.

File: server/channels/store/sqlstore/store.go (hunk 6 above)
Comment (davidkrauser) [reply server/channels/store/sqlstore/store.go:1283]: Could we check if `db_migrations` has the correct value for `add_roles_schema_index` instead? Maybe this could be wrapped in a function like `handler` (but to check if the migration has been applied).

**Review by davidkrauser — APPROVED** [review]
This looks great. Thank you again for cleaning up my mess 🙏

File: server/channels/store/sqlstore/store.go (hunk 6 above)
Comment (harshilsharma63) [reply server/channels/store/sqlstore/store.go:1283]: @davidkrauser We could, but I'm just keeping the identical pattern as app migrations for uniformity. No preference though.

**Review by isacikgoz — APPROVED** [review]

File: server/channels/store/sqlstore/store.go (hunk 3 above)
Comment (agarciamontoro) [reply server/channels/store/sqlstore/store.go:1193]: Not a fan of adding code just-in-case for the future, but 2/5, up to you to keep it or remove it.

File: server/channels/store/sqlstore/store.go [hunk 9]
```diff
@@ -1138,3 +1143,144 @@ func (ss *SqlStore) ScheduledPost() store.ScheduledPostStore {
 func (ss *SqlStore) ContentFlagging() store.ContentFlaggingStore {
 	return ss.stores.ContentFlagging
 }
+
+func (ss *SqlStore) preMigration() error {
```
Comment (agarciamontoro) [inline server/channels/store/sqlstore/store.go:1153]: Just another nit: can we specify what the goal of this `preMigration` function is as a comment? I'll find it weird that we even have a pre-migration six months from now when I've forgotten about this.

File: server/channels/store/sqlstore/store_test.go [hunk 10]
```diff
@@ -1105,3 +1105,201 @@ func TestSkipMigrationsOption(t *testing.T) {
...
+		"SELECT Version, Name FROM db_migrations WHERE Version IN (142, 143, 144, 156, 157, 158) ORDER BY Version"))
+
+	assert.Equal(t, []row{
+		{156, "add_schemeid_to_roles"},
+		{157, "backfill_roles_schemeid"},
+		{158, "add_roles_schemeid_index"},
+	}, rows)
+
+	// Running again is a safe no-op: WHERE clauses match nothing now.
+	require.NoError(t, ss.doRenumberRolesSchemeIdMigrations())
```
Comment (agarciamontoro) [inline server/channels/store/sqlstore/store_test.go:1219]: I would double-check after this that the `SELECT Version, Name FROM db_migrations WHERE Version IN (142, 143, 144, 156, 157, 158) ORDER BY Version` from above still returns the same.

**Review by agarciamontoro — CHANGES_REQUESTED** [review]
A couple more nits, but this looks good to me!

Two general comments:

1. I'm thinking about this running in HA, and although there is a theoretical race in the chain `isPreMigrationComplete` > run the pre-migration `m.handler()` > `markPreMigrationComplete`, the idempotency of every query involved makes it a non-issue. Just surfacing in case I'm missing something.
2. And I think this has been already discussed, but we need to cherry-pick this to every supported branch out there, right? Otherwise, an installation that is at the faulty version (the one with the renamed mentions), that is updated to a version that does not contain this fix would still fail to update.

File: server/channels/store/sqlstore/store_test.go (hunk 10 above)
Comment (harshilsharma63) [reply server/channels/store/sqlstore/store_test.go:1219]: Done.

File: server/channels/store/sqlstore/store.go (hunk 9 above)
Comment (harshilsharma63) [reply server/channels/store/sqlstore/store.go:1153]: Done.

File: server/channels/store/sqlstore/store.go [hunk 11]
```diff
@@ -1144,6 +1144,10 @@ func (ss *SqlStore) ContentFlagging() store.ContentFlaggingStore {
 	return ss.stores.ContentFlagging
 }
 
+// preMigration	runs before running the actual Morph migrations.
+// The main reason for having a preMigration is to fix https://mattermost.atlassian.net/browse/MM-68848?focusedCommentId=217769
+// However, using that as an opportunity, a general purpose system is created that allows running
+// arbitrary code just before the Morph migrations run. This allows us to have a way to fix any issues in the DB that might prevent the Morph migrations from running successfully. The pre-migrations are tracked in the Systems table to make sure they run only once.
```
Comment (agarciamontoro) [inline server/channels/store/sqlstore/store.go]: That's a loooong line.

**Review by agarciamontoro — APPROVED** [review]

File: server/channels/store/sqlstore/store.go (hunk 11 above)
Comment (harshilsharma63) [reply server/channels/store/sqlstore/store.go]: Done.

**Comment by harshilsharma63** [comment]
@amyblais we need to cherrypick this to all supported versions except the v10.11. So I believe those would be v11.5, v11.6, v11.7 and the upcoming v11.8?

**Comment by amyblais** [comment]
> we need to cherrypick this to all supported versions except the v10.11. So I believe those would be v11.5, v11.6, v11.7 and the upcoming v11.8?

@harshilsharma63 That's correct 👍

**Comment by harshilsharma63** [comment]
Hi @lindalumitchell this is ready for QA reviews. The most important thing to test here is an upgrade from v10.11 to this PR without destroying the database.  Please LMK if I can help with this.

**Comment by lindalumitchell** [comment]
@harshilsharma63 I prepared a test server with lots of seeded data, permissions scheme, etc., but the upgrade is hanging. 

I currently have two cloud servers attempting to update to this PR version, one starting from v10.11.19 and one (just to compare) starting from v11.8, but neither is completing. I used what I typically use, `/cloud update lmmigrate --image mattermostdevelopment/mattermost-enterprise-edition --version 670a60c`. Not sure why it wouldn't be working this time. 

Maybe next week we can look at this together?

**Comment by harshilsharma63** [comment]
@lindalumitchell No seed data is needed to test this upgrade path. An empty server is sufficient to reproduce this bug, and test the upgrade.

Do you know where can I access the logs for the upgrade you're performing to see whats wrong?

**Comment by lindalumitchell** [comment]
@harshilsharma63: 

> No seed data is needed to test this upgrade path. An empty server is sufficient to reproduce this bug, and test the upgrade.

I'll try again, specifying the image from the start instead of just during the upgrade. Easier to quickly give that a try with an empty server.

[Installation logs are here](https://grafana.internal.mattermost.com/explore?schemaVersion=1&panes=%7B%222fn%22%3A%7B%22datasource%22%3A%22PFB2D5CACEC34D62E%22%2C%22queries%22%3A%5B%7B%22refId%22%3A%22A%22%2C%22datasource%22%3A%7B%22type%22%3A%22loki%22%2C%22uid%22%3A%22PFB2D5CACEC34D62E%22%7D%2C%22editorMode%22%3A%22code%22%2C%22expr%22%3A%22%7Bapp%3D%5C%22mattermost%5C%22%2C+namespace%3D%5C%22p3jbpd556fygmn9edmno9oxjuw%5C%22%7D%22%2C%22queryType%22%3A%22range%22%2C%22direction%22%3A%22backward%22%7D%5D%2C%22range%22%3A%7B%22from%22%3A%22now-1h%22%2C%22to%22%3A%22now%22%7D%7D%7D&orgId=1) for the `lmmigrate` server I tried to update -- not sure if that's helpful, but that's all I can get from the cloud plugin.

**Comment by lindalumitchell** [comment]
For visibility and reference, Harshil and I have been chatting about what I'm seeing, and I shared the following chart of scenarios and results I've tried so far:

> I tried a ton of different scenarios and some are stuck and some worked. I can't make sense of which did which. 

> In my mind, it should fail / get stuck if I try to update v10.11.17 to v11.7 (as reported in the ticket), ~~but that update worked fine~~ EDIT: which it did, but then updating the same v10.11.17 version to the PR build failed when I expected it to work. :/

> Lots of updates to the PR build got stuck, but not all of them. So the update method isn't faulty.

> Here's a table I just finished making, that shows most of the tests I ran. You can see I tried various version types including release versions or RCs or final versions. It's a lot. I'm happy to point out important parts or walk you through it if it's impossible to sort out.

| pass/fail | starting version | updated to version | server name | notes | installation logs |
| --- | --- | --- | --- | --- | --- |
| :x: | 10.11.17 (202d125_7893d97) | 11.7  (a9e574a_35a120e) | lm101117to117 | 🎉 failed as expected | [logs](https://grafana.internal.mattermost.com/explore?schemaVersion=1&panes=%7B%22l59%22%3A%7B%22datasource%22%3A%22PFB2D5CACEC34D62E%22%2C%22queries%22%3A%5B%7B%22refId%22%3A%22A%22%2C%22datasource%22%3A%7B%22type%22%3A%22loki%22%2C%22uid%22%3A%22PFB2D5CACEC34D62E%22%7D%2C%22editorMode%22%3A%22code%22%2C%22expr%22%3A%22%7Bapp%3D%5C%22mattermost%5C%22%2C+namespace%3D%5C%221a8bw6kmuig73n7z8wtkpea3oa%5C%22%7D%22%2C%22queryType%22%3A%22range%22%2C%22direction%22%3A%22backward%22%7D%5D%2C%22range%22%3A%7B%22from%22%3A%22now-1h%22%2C%22to%22%3A%22now%22%7D%7D%7D&orgId=1) |
| :x: | 10.11.17 (202d125_7893d97) | PR build  (670a60c) | lm101117to-pr | expected to work, but failed | [logs](https://grafana.internal.mattermost.com/explore?schemaVersion=1&panes=%7B%22jo9%22%3A%7B%22datasource%22%3A%22PFB2D5CACEC34D62E%22%2C%22queries%22%3A%5B%7B%22refId%22%3A%22A%22%2C%22datasource%22%3A%7B%22type%22%3A%22loki%22%2C%22uid%22%3A%22PFB2D5CACEC34D62E%22%7D%2C%22editorMode%22%3A%22code%22%2C%22expr%22%3A%22%7Bapp%3D%5C%22mattermost%5C%22%2C+namespace%3D%5C%227miaacfrktgubrmz366dwjhyqw%5C%22%7D%22%2C%22queryType%22%3A%22range%22%2C%22direction%22%3A%22backward%22%7D%5D%2C%22range%22%3A%7B%22from%22%3A%22now-1h%22%2C%22to%22%3A%22now%22%7D%7D%7D&orgId=1) |
| :x: | 10.11.19-rc1 () | PR build  (670a60c) | lm101119-rc1 | also expected to work, but failed | [logs](https://grafana.internal.mattermost.com/explore?schemaVersion=1&panes=%7B%224qp%22%3A%7B%22datasource%22%3A%22PFB2D5CACEC34D62E%22%2C%22queries%22%3A%5B%7B%22refId%22%3A%22A%22%2C%22datasource%22%3A%7B%22type%22%3A%22loki%22%2C%22uid%22%3A%22PFB2D5CACEC34D62E%22%7D%2C%22editorMode%22%3A%22code%22%2C%22expr%22%3A%22%7Bapp%3D%5C%22mattermost%5C%22%2C+namespace%3D%5C%22saxm7kxaoff78mkgsw7uispgfr%5C%22%7D%22%2C%22queryType%22%3A%22range%22%2C%22direction%22%3A%22backward%22%7D%5D%2C%22range%22%3A%7B%22from%22%3A%22now-1h%22%2C%22to%22%3A%22now%22%7D%7D%7D&orgId=1) |
| :x: | 10.11.19 (10.11.19) | PR build  (670a60c)  | lmmigrate2 | also expected to work, but failed | [logs](https://grafana.internal.mattermost.com/explore?schemaVersion=1&panes=%7B%223b8%22%3A%7B%22datasource%22%3A%22PFB2D5CACEC34D62E%22%2C%22queries%22%3A%5B%7B%22refId%22%3A%22A%22%2C%22datasource%22%3A%7B%22type%22%3A%22loki%22%2C%22uid%22%3A%22PFB2D5CACEC34D62E%22%7D%2C%22editorMode%22%3A%22code%22%2C%22expr%22%3A%22%7Bapp%3D%5C%22mattermost%5C%22%2C+namespace%3D%5C%224pf8iz7ugjd6imke5fu4gfsj3y%5C%22%7D%22%2C%22queryType%22%3A%22range%22%2C%22direction%22%3A%22backward%22%7D%5D%2C%22range%22%3A%7B%22from%22%3A%22now-1h%22%2C%22to%22%3A%22now%22%7D%7D%7D&orgId=1) |
| :x: | release-10.11 (release-10.11) | PR build  (670a60c)  | lmmigrate3 | also expected to work, but failed | [logs](https://grafana.internal.mattermost.com/explore?schemaVersion=1&panes=%7B%22xmd%22%3A%7B%22datasource%22%3A%22PFB2D5CACEC34D62E%22%2C%22queries%22%3A%5B%7B%22refId%22%3A%22A%22%2C%22datasource%22%3A%7B%22type%22%3A%22loki%22%2C%22uid%22%3A%22PFB2D5CACEC34D62E%22%7D%2C%22editorMode%22%3A%22code%22%2C%22expr%22%3A%22%7Bapp%3D%5C%22mattermost%5C%22%2C+namespace%3D%5C%22haka67ppapd87dad1r6zzet67y%5C%22%7D%22%2C%22queryType%22%3A%22range%22%2C%22direction%22%3A%22backward%22%7D%5D%2C%22range%22%3A%7B%22from%22%3A%22now-1h%22%2C%22to%22%3A%22now%22%7D%7D%7D&orgId=1) |
| :x: | release-11.8 |  PR build  (670a60c)  | lm118release | I think I expected this to work? | [logs](https://grafana.internal.mattermost.com/explore?schemaVersion=1&panes=%7B%221cn%22%3A%7B%22datasource%22%3A%22PFB2D5CACEC34D62E%22%2C%22queries%22%3A%5B%7B%22refId%22%3A%22A%22%2C%22datasource%22%3A%7B%22type%22%3A%22loki%22%2C%22uid%22%3A%22PFB2D5CACEC34D62E%22%7D%2C%22editorMode%22%3A%22code%22%2C%22expr%22%3A%22%7Bapp%3D%5C%22mattermost%5C%22%2C+namespace%3D%5C%22ei74hs8igbf3bd3bbyexx51cir%5C%22%7D%22%2C%22queryType%22%3A%22range%22%2C%22direction%22%3A%22backward%22%7D%5D%2C%22range%22%3A%7B%22from%22%3A%22now-1h%22%2C%22to%22%3A%22now%22%7D%7D%7D&orgId=1) |
| ✅  | 10.11.0 (5b95546_93db4e1) | 11.7  (a9e574a_35a120e) | lm10110rc3to117 |  worked as expected | [logs](https://grafana.internal.mattermost.com/explore?schemaVersion=1&panes=%7B%226m6%22%3A%7B%22datasource%22%3A%22PFB2D5CACEC34D62E%22%2C%22queries%22%3A%5B%7B%22refId%22%3A%22A%22%2C%22datasource%22%3A%7B%22type%22%3A%22loki%22%2C%22uid%22%3A%22PFB2D5CACEC34D62E%22%7D%2C%22editorMode%22%3A%22code%22%2C%22expr%22%3A%22%7Bapp%3D%5C%22mattermost%5C%22%2C+namespace%3D%5C%22ei74hs8igbf3bd3bbyexx51cir%5C%22%7D%22%2C%22queryType%22%3A%22range%22%2C%22direction%22%3A%22backward%22%7D%5D%2C%22range%22%3A%7B%22from%22%3A%22now-1h%22%2C%22to%22%3A%22now%22%7D%7D%7D&orgId=1) |
| ✅ | release-11.4 (release-11.4) | PR build  (670a60c) | lm114 | ? | [logs](https://grafana.internal.mattermost.com/explore?schemaVersion=1&panes=%7B%220mm%22%3A%7B%22datasource%22%3A%22PFB2D5CACEC34D62E%22%2C%22queries%22%3A%5B%7B%22refId%22%3A%22A%22%2C%22datasource%22%3A%7B%22type%22%3A%22loki%22%2C%22uid%22%3A%22PFB2D5CACEC34D62E%22%7D%2C%22editorMode%22%3A%22code%22%2C%22expr%22%3A%22%7Bapp%3D%5C%22mattermost%5C%22%2C+namespace%3D%5C%22beeuoyyq73f138sy51m9b9ghsr%5C%22%7D%22%2C%22queryType%22%3A%22range%22%2C%22direction%22%3A%22backward%22%7D%5D%2C%22range%22%3A%7B%22from%22%3A%22now-1h%22%2C%22to%22%3A%22now%22%7D%7D%7D&orgId=1) |
| ✅  | release-10.10 (release-10.10) | PR build  (670a60c) | lm1010 | worked as expected | [logs](https://grafana.internal.mattermost.com/explore?schemaVersion=1&panes=%7B%22y9a%22%3A%7B%22datasource%22%3A%22PFB2D5CACEC34D62E%22%2C%22queries%22%3A%5B%7B%22refId%22%3A%22A%22%2C%22datasource%22%3A%7B%22type%22%3A%22loki%22%2C%22uid%22%3A%22PFB2D5CACEC34D62E%22%7D%2C%22editorMode%22%3A%22code%22%2C%22expr%22%3A%22%7Bapp%3D%5C%22mattermost%5C%22%2C+namespace%3D%5C%22k51y4n7jmfbadgbdgy7r6ngixw%5C%22%7D%22%2C%22queryType%22%3A%22range%22%2C%22direction%22%3A%22backward%22%7D%5D%2C%22range%22%3A%7B%22from%22%3A%22now-1h%22%2C%22to%22%3A%22now%22%7D%7D%7D&orgId=1)

File: server/channels/store/sqlstore/migrate.go [hunk 12]
```diff
@@ -172,6 +172,19 @@ func (m *Migrator) MigrateWithPlan(plan *models.Plan, dryRun bool) error {
 	return m.engine.ApplyPlan(plan)
 }
 
+// PreMigrate runs the pre-migration handlers that normally execute during
+// server startup. Callers on the CLI up-migration path (e.g. `mattermost db
+// migrate`, used by cloud upgrades) must invoke this before MigrateWithPlan so
+// the same fixes apply outside of server startup. Skipped under dryRun because
+// preMigration writes directly via GetMaster().Exec and does not participate
+// in Morph's dry-run. Must NOT be called from the downgrade path.
```
Comment (agarciamontoro) [inline server/channels/store/sqlstore/migrate.go]: > Must NOT be called from the downgrade path.

Is there a way to prevent this with code?

**Review by agarciamontoro — COMMENTED** [review]
Just a question below

File: server/channels/store/sqlstore/migrate.go (hunk 12 above)
Comment (harshilsharma63) [reply server/channels/store/sqlstore/migrate.go]: I believe the code comment is now invalid as I previously had it in a place that could be called from both upgrade and downgrade, but now I'm only calling from the upgrade function. So, I can update the code comment to say that it is intentional to only have this called in upgrade path.

**Review by lindalumitchell — APPROVED** [review]
The fix looks great! I ran through a set of upgrades, and all worked (or failed) as expected. LGTM!

| pass/fail | starting version | updated to version | server name | notes | installation logs |
| --- | --- | --- | --- | --- | --- |
| :white_check_mark: | 10.11.17 (202d125_7893d97) | PR build (44c860b) | fix-lm101117to-pr | 🎉 worked as expected | |
| :white_check_mark: | 10.11.17 (202d125_7893d97) | 11.7  (a9e574a_35a120e) | fix-lm101117to117 | 🎉  failed as expected | |
| :white_check_mark: | 10.11.19 (4439c61_7893d97) | PR build (44c860b) | fix-lm101119to-pr | 🎉  worked as expected | |
| :white_check_mark: | 10.11.19 (202d125_7893d97) | PR build (44c860b) | fixdata-lm101117to-pr | 🎉 worked as expected, with seed data | |

I also ran an upgrade with seed data based on Matty's recommendations and using a script it generated. Attaching the script and results for reference. All passed.

[mm_migration_test.py](https://github.com/user-attachments/files/28764782/mm_migration_test.py)

[pre_upgrade_roles.txt](https://github.com/user-attachments/files/28764788/pre_upgrade_roles.txt)
[pre_upgrade_snapshot.txt](https://github.com/user-attachments/files/28764789/pre_upgrade_snapshot.txt)

[post_upgrade_roles.txt](https://github.com/user-attachments/files/28764797/post_upgrade_roles.txt)
[post_upgrade_snapshot.txt](https://github.com/user-attachments/files/28764796/post_upgrade_snapshot.txt)

[diff_snapshots.txt](https://github.com/user-attachments/files/28764802/diff_snapshots.txt)
[diff_upgrade_roles.txt](https://github.com/user-attachments/files/28764801/diff_upgrade_roles.txt)

File: server/cmd/mattermost/commands/db.go [hunk 13]
```diff
@@ -216,6 +216,10 @@ func migrateCmdF(command *cobra.Command, args []string) error {
 				strings.Repeat("*", 80), fileName+".json", strings.Repeat("*", 80)))
 	}
 
+	if preMigrationError := migrator.PreMigrate(dryRun); preMigrationError != nil {
```
Comment (davidkrauser) [inline server/cmd/mattermost/commands/db.go]: **suggestion (non-blocking):** if dryRun is set, should we just not call `PreMigrate()` at all? Then you don't need to add a parameter to `PreMigrate()`.

**Review by davidkrauser — APPROVED** [review]

File: server/cmd/mattermost/commands/db.go (hunk 13 above)
Comment (harshilsharma63) [reply server/cmd/mattermost/commands/db.go]: Ah, thats such a simple solution. Done!

**Comment by lindalumitchell** [comment]
Just to note that I ran a quick test with the latest commit and it still updates as expected; no issues found. :)
